### DISCUSSION TITLE
release cadence

### DISCUSSION THREAD

**Discussion by mil-ad** [ahrm/sioyek#1156] [Q&A] [answered]
I know the dev branch is quite active and a lot of work is going into iOS and android versions (presumbaly for a 3.0 release) but it would be great if we could get more regular 2.x.x releases. The last official release is from more than two years ago.

**Comment by ahrm** [comment]
I have created a preview build of the development branch here: https://github.com/ahrm/sioyek/releases/tag/sioyek3-alpha0 . I have not tested the binaries on macos or windows though, it is quite possible that they don't work.

**Comment by mil-ad** [comment]
Thanks for this @ahrm.

It seems to work for me on macOS. I'm getting some warnings when starting the app but not sure if they matter or not:

```
default_config_path: /Applications/sioyek.app/Contents/MacOS/prefs.config
default_keys_path: /Applications/sioyek.app/Contents/MacOS/keys.config
user_config_path: [ 0 ] /Users/milad/.config/sioyek/prefs_user.config
user_config_path: [ 1 ] /Users/milad/Library/Application Support/sioyek/prefs_user.config
user_keys_path: [ 0 ] /Users/milad/.config/sioyek/keys_user.config
user_keys_path: [ 1 ] /Users/milad/Library/Application Support/sioyek/keys_user.config
database_file_path: /Users/milad/Library/Application Support/sioyek/test.db
local_database_file_path: /Users/milad/Library/Application Support/sioyek/local.db
global_database_file_path: /home/milad/Seafile/documents/sioyek_database.db
tutorial_path: /Applications/sioyek.app/Contents/MacOS/tutorial.pdf
last_opened_file_address_path: /Users/milad/Library/Application Support/sioyek/last_document_path.txt
shader_path: /Applications/sioyek.app/Contents/MacOS/shaders
could not create global databaseunable to open database file
SQL Error in  "get_version"  :   

Error: Invalid database version
SIOYEK
Warning: key defined in /Applications/sioyek.app/Contents/MacOS/keys.config:288 overwritten by /Users/milad/Library/Application Support/sioyek/keys_user.config:1. Overriding command: q: replacing quit with close_window
Warning: key defined in /Applications/sioyek.app/Contents/MacOS/keys.config:285 overwritten by /Users/milad/Library/Application Support/sioyek/keys_user.config:2. Overriding command: <f5>: replacing toggle_presentation_mode with reload
Warning: key defined in /Applications/sioyek.app/Contents/MacOS/keys.config:101 overwritten by /Users/milad/Library/Application Support/sioyek/keys_user.config:3. Overriding command: r: replacing rotate_clockwise with reload
Warning: key defined in /Applications/sioyek.app/Contents/MacOS/keys.config:291 overwritten by /Users/milad/Library/Application Support/sioyek/keys_user.config:4. Overriding command: f: replacing open_link with toggle_fullscreen
Warning: key defined in /Applications/sioyek.app/Contents/MacOS/keys.config:227 overwritten by /Users/milad/Library/Application Support/sioyek/keys_user.config:5. Overriding command: <tab>: replacing goto_portal with goto_toc
Warning: key defined in /Applications/sioyek.app/Contents/MacOS/keys.config:281 overwritten by /Users/milad/Library/Application Support/sioyek/keys_user.config:7. Overriding command: <C-]>: replacing goto_definition with next_state
Warning: key defined in /Applications/sioyek.app/Contents/MacOS/keys.config:255 overwritten by /Users/milad/Library/Application Support/sioyek/keys_user.config:12. Overriding command: s: replacing external_search with fit_to_page_height
Warning: key defined in /Applications/sioyek.app/Contents/MacOS/keys.config:23 overwritten by /Users/milad/Library/Application Support/sioyek/keys_user.config:15. Overriding command: gg: replacing goto_beginning with goto_beginning;goto_top_of_page
Warning: key defined in /Applications/sioyek.app/Contents/MacOS/keys.config:31 overwritten by /Users/milad/Library/Application Support/sioyek/keys_user.config:16. Overriding command: <home>: replacing goto_page_with_page_number with goto_beginning;goto_top_of_page
Warning: key defined in /Applications/sioyek.app/Contents/MacOS/keys.config:76 overwritten by /Users/milad/Library/Application Support/sioyek/keys_user.config:17. Overriding command: <C-t>: replacing new_window with toggle_custom_color
SQL Error in  "get_prev_path_hash_pairs"  :   

WARNING: Output of vertex shader 'screen_pos' not read by fragment shader

WARNING: Output of vertex shader 'screen_pos' not read by fragment shader

WARNING: Output of vertex shader 'screen_pos' not read by fragment shader

WARNING: Output of vertex shader 'uvs' not read by fragment shader
WARNING: Output of vertex shader 'screen_pos' not read by fragment shader

WARNING: Output of vertex shader 'uvs' not read by fragment shader
WARNING: Output of vertex shader 'screen_pos' not read by fragment shader

WARNING: Output of vertex shader 'uvs' not read by fragment shader
WARNING: Output of vertex shader 'screen_pos' not read by fragment shader

WARNING: Output of vertex shader 'screen_pos' not read by fragment shader

WARNING: Output of vertex shader 'uvs' not read by fragment shader
WARNING: Output of vertex shader 'screen_pos' not read by fragment shader

WARNING: Output of vertex shader 'screen_pos' not read by fragment shader

WARNING: 0:2: extension 'GL_OES_standard_derivatives' is not supported

SQL Error in  "insert_document_hash"  :   

SQL Error in  "insert_document_hash"  :   
SQL Error in  "update_book"  :   
```

**Comment by mil-ad** [comment]
We (realistically you but I'm happy to help if needed) should probably clarify if the macOS build is for x86 or Apple silicon (i.e. ARM64) and release two versions?

**Comment by ahrm** [comment]
The SQL errors do matter, this preview release is not compatible with the old database. I recommend moving your `local.db` and `shared.db` file to another location and see if the errors still appear (I could write a script to migrate the database to the new version if you want).

> We (realistically you but I'm happy to help if needed) should probably clarify if the macOS build is for x86 or Apple silicon (i.e. ARM64) and release two versions?

Isn't this already the case? We have a `sioyek-release-mac` build which should be x64 and a `sioyek-release-mac-arm` build which should be arm64.

**Reply by jpaulogg** [reply]
> (I could write a script to migrate the database to the new version if you want).

I would really appreciate that script! I'm currently experiencing issues with the highlights, similar to the ones mentioned in issue #583

**Reply by ahrm** [reply]
I have already posted the script below: https://github.com/ahrm/sioyek/discussions/1156#discussioncomment-10614340.

**Comment by mil-ad** [comment]
> Isn't this already the case? We have a sioyek-release-mac build which should be x64 and a sioyek-release-mac-arm build which should be arm64.

You're right. I'm blind. Ignore me please

**Comment by mil-ad** [comment]
> I could write a script to migrate the database to the new version if you want).

That'd probably be useful for a lot of people

**Reply by ahrm** [reply]
I wrote a script to upgrade the database here: https://gist.github.com/ahrm/c402a1664c769ac3a113901355983922.

Here is how to use it:
1. Backup your database files
2. Delete the old `shared.db` file so that sioyek creates a new one with the correct structure.
3. Run the script like so: `python database_migrator.py --old-shared-db /path/to/old/shared.db --new-shared-db /path/to/sioyek/generated/shared.db`

Let me know if there are any issues.

**Reply by Vortriz** [reply]
I am getting the following error:

```
Migrating table opened_books
Traceback (most recent call last):
  File "D:\downloads\programs\sioyek backup\c402a1664c769ac3a113901355983922-fc816b6f52aebe5a37186dd8c9187e8a44403e20\database_migrator.py", line 109, in <module>
    migrate(old_conn, new_conn)
  File "D:\downloads\programs\sioyek backup\c402a1664c769ac3a113901355983922-fc816b6f52aebe5a37186dd8c9187e8a44403e20\database_migrator.py", line 96, in migrate
    migrate_table(old_conn, new_conn, table)
  File "D:\downloads\programs\sioyek backup\c402a1664c769ac3a113901355983922-fc816b6f52aebe5a37186dd8c9187e8a44403e20\database_migrator.py", line 84, in migrate_table
    cursor.execute(f"INSERT INTO {table_name} ({columns}) VALUES ({values})")
sqlite3.IntegrityError: UNIQUE constraint failed: opened_books.path
```

**Reply by ahrm** [reply]
Can you share an example of a file that causes this error?

**Reply by Vortriz** [reply]
Here are the old and new db files.
[sioyek db.zip](https://github.com/user-attachments/files/17019644/sioyek.db.zip)


lemme know if you need anything else.

**Reply by ahrm** [reply]
I have updated the code, let me know if it still has issues.

**Reply by Vortriz** [reply]
works perfectly!

**Reply by miss4chewN8** [reply]
> I have updated the code, let me know if it still has issues.

@ahrm it works perfectly on macOS too.

**Reply by svonjoi** [reply]
hi, i cant migrate highlights

```
Migrating table opened_books
Migrated 31 rows from opened_books
Migrating table marks
Migrated 11 rows from marks
Migrating table bookmarks
Migrated 23 rows from bookmarks
Migrating table highlights
Traceback (most recent call last):
  File "/tmp/script2.py", line 109, in <module>
    migrate(old_conn, new_conn)
    ~~~~~~~^^^^^^^^^^^^^^^^^^^^
  File "/tmp/script2.py", line 96, in migrate
    migrate_table(old_conn, new_conn, table)
    ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/script2.py", line 84, in migrate_table
    cursor.execute(f"INSERT OR REPLACE INTO {table_name} ({columns}) VALUES ({values})")
    ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
sqlite3.OperationalError: near "s": syntax error
```

**Reply by svonjoi** [reply]
> hi, i cant migrate highlights
> 
> ```
> Migrating table opened_books
> Migrated 31 rows from opened_books
> Migrating table marks
> Migrated 11 rows from marks
> Migrating table bookmarks
> Migrated 23 rows from bookmarks
> Migrating table highlights
> Traceback (most recent call last):
>   File "/tmp/script2.py", line 109, in <module>
>     migrate(old_conn, new_conn)
>     ~~~~~~~^^^^^^^^^^^^^^^^^^^^
>   File "/tmp/script2.py", line 96, in migrate
>     migrate_table(old_conn, new_conn, table)
>     ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/tmp/script2.py", line 84, in migrate_table
>     cursor.execute(f"INSERT OR REPLACE INTO {table_name} ({columns}) VALUES ({values})")
>     ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> sqlite3.OperationalError: near "s": syntax error
> ```

the problem are the commas and single quotes in hightlight, bookmark, etc entries. Single quotes dont being escaped in the insert statment and the comma "," character destroy the insert builder itself cuz it relies on comma to separate values by columns

here is a shitpatch to solve this by removing this characters in the migration. Replace the 2 lines [here](https://gist.github.com/ahrm/c402a1664c769ac3a113901355983922#file-database_migrator-py-L71)

```python
    for row in rows:
        old_values.append({old_columns[i][0]: row[i] for i in range(len(row))})
```

with these lines

```python
    for row in rows:
        data = {old_columns[i][0]: row[i] for i in range(len(row))}
        # remove "’", "'", "," from "desc" column before insert
        if any(char in data.get("desc", "") for char in ["’", "'", ","]):
            print("<<<BEFORE: " + data.get("desc"))
            for char in ["’", "'", ","]:
                data["desc"] = data["desc"].replace(char,"")
            print(">>>AFTER: " + data.get("desc"))
        old_values.append(data)
```

**Comment by bragiato** [comment]
It seems to work for me on Windows! Thx!

**Comment by mil-ad** [comment]
on macOS I get the below when I try to run the application:

<img width="372" alt="image" src="https://github.com/user-attachments/assets/b305550a-87f7-465a-a9f5-4191a73d459b">

(I'm on Secoma 14.6 (23G80) / M3 Max)

**Reply by ahrm** [reply]
I don't have a mac so I can't test it but I assume it might be a permission issue? You might want to go to System Preference > Privacy & Security and click on open anyway after trying to launch sioyek.

**Reply by mil-ad** [reply]
unfortunately it's not that :( 

However, the x86 version does work (on ARM with Rosetta translation) so I think it's something with the arm build specifically

**Reply by miss4chewN8** [reply, marked as answer]
> on macOS I get the below when I try to run the application:
> 
> <img alt="image" width="372" src="https://private-user-images.githubusercontent.com/686449/372806290-b305550a-87f7-465a-a9f5-4191a73d459b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc4ODQ4MzMsIm5iZiI6MTcyNzg4NDUzMywicGF0aCI6Ii82ODY0NDkvMzcyODA2MjkwLWIzMDU1NTBhLTg3ZjctNDY1YS1hOWY1LTQxOTFhNzNkNDU5Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAwMlQxNTU1MzNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03NDliMWNkMjE3YTFiNzc4YTc3OGI2ZTM4MjAzMDkyOTBiZWZlNTIzNGY3MzBhZGU2OWU5NjFiZWIzZDJjMWQ0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.uNjWTtoZOMsHZynBEVozKooC8zFxawO0C6F3urU--u4">
> (I'm on Secoma 14.6 (23G80) / M3 Max)

@mil-ad This is related to macOS quarantine. You need to run the following command on your terminal:

`xattr -c /Applications/sioyek.app`

and then it should run properly. As to why this is not a problem for the x86 build, I am not sure but after doing this it worked perfectly for me.

**Reply by mil-ad** [reply]
that worked. Thanks!

**Comment by gitsroid** [comment]
@ahrm Could you pin this discussion? Currently there are many discussions about the same and it is annoying. Maybe pinning it could help.

**Comment by jweisber** [comment]
Anybody else having trouble with inverse search on macOS in this release? It works out of the box for me on 2.0.0 but not on this release. I've tried setting `inverse_search_command` in `prefs_user.config`, but no joy.
